### DISCUSSION TITLE
Piecewise Linear Regression - family for non-normal distribution with excess of  zero’s

### DISCUSSION THREAD

**Discussion by lamaenho** [jazznbass/scan#139] [Q&A] [answered]
Within the Scan package, I am using the plm() function for conducting a piecewise linear regression. However, my outcome variable (moderate-to-vigorous physical activity) is not normally distributed, so I can't use the 'gaussian' family. When trying another distribution, both the inverse.gaussian as the gamma distributions are not working, because I have a lot of zeros (non-positive values). Both 'families' are providing erros in R, as positive values only are allowed for the 'inverse.gaussian' family and non-positive values not allowed for the 'Gamma' family. Is there somebody that can provide help, please? :-)

**Comment by jazznbass** [comment]
I would need more information on the scale of the dependent variable. Has the scale a natural zero? Does it depict frequencies or proportions? If so, a negative binomial or poisson distribution might be appropriate. 
If not, you could transform the values from a gamma to a nornal distribution if they are really gamma distributed.

**Reply by lamaenho** [reply]
My dependent variable (physical activity) is measured in minutes/day, which is considered to have a natural zero (a natural zero would imply complete inactivity or the absence of any physical activity). It does not depict frequencies or proportions. I did a Shapiro-Wilk normality test, showing my dependent variable is not normal, and the histogram shows a right skewed distribution. 

The following command gives me the error that non-positive values (so zeros) are not allowed for the 'Gamma' family:

_plm (scedPA_002_imptd, dvar = "imptd_MVPA_bts10", pvar = "Phase", mvar = "prog_day", model = "W", family= "Gamma", trend = TRUE, level = TRUE, slope = TRUE, contrast_level = "preceding", contrast_slope = "preceding")_

Any advice how I can transform the values from a gamma to a normal distribution, using a single case dataframe? 
(_scedPA_002_imptd <- readSC("P002_imptd.csv", cvar = "case", dvar = "imptd_MVPA_bts10", pvar = "Phase", mvar = "prog_day")_)

Thanks for the quick response and help!

**Reply by jazznbass** [reply]
Here is the code for an example that transforms gamma distributed data to normal distributed:

```{r}
library(scan)

set.seed(123)

# create an example-case with gamma distributed data
case <- scdf(
  c(rgamma(40, 5), rgamma(40, 10)), 
  phase_design = list(A = 40, B = 40)
)

# create a transformed variable and use the gamma2norm function from the faux package (install it from cran). 
# Mean and sd for the transformation are estimated from the sample data. 
case <- case |> transform(
  values_trans = faux::gamma2norm(
    values, 
    mu = all_cases(mean(values)), 
    sd = all_cases(sd(values)))
)

# Check with shapiro original and transformed distributions for deviation from normality
shapiro.test(case[[1]]$values)
shapiro.test(case[[1]]$values_trans)

# Example plm analysis
plm(case, dvar = "values_trans")
```

**Reply by lamaenho** [reply]
Thanks again for the help. Running the R code as provided above, still shows me a problem with the days that participants had 0 minutes of physical activity, making it impossible to run the command: 
It seems like there might be an issue with the "values_trans" variable, which may contain missing values (NA), NaN (Not a Number), or Inf (Infinity). The error in the plm function suggests that there are problematic values in the response variable.

The same happens when I try it another way: 

_gamma_vector <- P001_imptd$imptd_MVPA_bts10
normal_vector <- gamma2norm(gamma_vector, mu = 38.45181, sd = 38.59228, shape = 38.4518072289157, rate = 0.1)
P001_imptd$imptd_MVPA_bts10_transformed <- normal_vector
hist(imptd_MVPA_bts10_transformed)

scedPA_001_imptd <- readSC("P001_imptd.csv", cvar = "case", dvar = "imptd_MVPA_bts10_transformed", pvar = "Phase", mvar = "prog_day")

plm_001_imptd <- plm (scedPA_001_imptd, dvar = "imptd_MVPA_bts10_transformed", pvar = "Phase", mvar = "prog_day", model = "W",
                      family = "gaussian",trend = TRUE,level = TRUE,slope = TRUE, contrast_level = "preceding", contrast_slope = "preceding")
plm_001_imptd_

**==> Error:  scdf has dependent variable 'imptd_MVPA_bts10_transformed' but not all cases have a variable named 'imptd_MVPA_bts10_transformed'.**

**Reply by jazznbass** [reply, marked as answer]
The error occurs when you are trying to predict infinite values.
If you have zero values than the data are not gamma distributed as this would imply an infinite low probability of such a value occurring.
You might try another transformation (people use sqrt transformations for seconds), try another method (randomization tests), or go with the non-normal distributed data as they are.

You could also ask on stackoverflow or crossvalidated, that is what I would go for.
